kernel: disable virtio mmio on arm64 for kvm hypervisor guest

justin-he opened the pull request.

jodh-intel commented on kernel/build-kernel.sh:
```diff
@@ -357,7 +360,11 @@ setup_kernel() {
 
-	[ -n "${hypervisor_target}" ] || hypervisor_target="kvm"
+	if [ -z "${hypervisor_target}" ]; then
+		info "hypervisor_target (-t) is not specified, use kvm instead"
```
Nit: s/use/using/.

justin-he replied:
ok

Pennyzct commented:
Firecracker really had very small sets of functionalities compared with qemu, like not supporting memory hotplug, cpu hotplug, etc. so maybe we should think about creating different kernel configs for different hypervisor type?

grahamwhaley commented:
> Firecracker really had very small sets of functionalities compared with qemu, like not supporting memory hotplug, cpu hotplug, etc. so maybe we should think about creating different kernel configs for different hypervisor type?

The idea makes sense to me, that we can/could have the minimal kernel necessary for each VM type - but, the downside is the complexity and bloat it will add to our configs and testing matrix....

/cc @kata-containers/architecture-committee for input here - I am thinking in particular around the probably move to CLH with kata v2.0 etc., as that will be in a similar-ish position to fc I think.

Pennyzct commented:
Hi~ @justin-he you may want to update kata_config_version. ;)

justin-he commented:
> Hi~ @justin-he you may want to update kata_config_version. ;)

ok ,updated config version in each single patch. Thanks

grahamwhaley commented:
@justin-he - you only need to change the kernel version in one patch/commit :-) - and, it has moved on already, so you'll need to rebase/repush that.
For the benchmark results - can you use the [metrics report tool](https://github.com/kata-containers/tests/tree/master/metrics/report) to get us a graph of the results over a number of runs - that way we can check if there is noise/variance, and not just a 'one off' win :-) thx!

justin-he commented:
> @justin-he - you only need to change the kernel version in one patch/commit :-) - and, it has moved on already, so you'll need to rebase/repush that.
ok
> For the benchmark results - can you use the [metrics report tool](https://github.com/kata-containers/tests/tree/master/metrics/report) to get us a graph of the results over a number of runs - that way we can check if there is noise/variance, and not just a 'one off' win :-) thx!


Ok, will do it later

justin-he committed:
kernel: arm64: split the kernel config for kvm and firecracker

There is some difference of the guest kernel between hypervisor kvm and fc.
E.g. on arm64, kvm hypervisor will use virtio pci bus instead of virtio
mmio bus. see the comparison from Richard Jones [1]

For kvm hypervisor guest on arm64, virtio mmio will be probed as the platform
device for 32 times even qemu hasn't created it. If we disable the virtio
mmio in kernel config, it will reduce 20-30ms for the boottime. But,
firecracker might only have the support for virtio mmio. Hence it
is worthy of splitting the kernel config into different. This patch simply
copy arm64_kata_kvm_5.4.x to arm64_kata_firecracker_5.4.x

[1] https://groups.google.com/a/groups.riscv.org/forum/#!topic/sw-dev/3MLIq6rxa9Q

Fixes: #985
Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
kernel: disable virtio mmio on arm64 for kvm hypervisor guest

To measure the performance improvement, see the timestamp of last dmesg:

=Before this patch=
[    0.391728] pci_bus 0000:01: busn_res: [bus 01] end is updated to 01
[    0.393129] pci 0000:00:02.0: PCI bridge to [bus 01]

=After this patch
[    0.367532] pci_bus 0000:01: busn_res: [bus 01] end is updated to 01
[    0.368919] pci 0000:00:02.0: PCI bridge to [bus 01]

It reduces about 25ms for boot time of kvm hypervisor guest on arm64

Fixes: #985
Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
kernel: arm64: fragments: Disable virtio mmio device in arm64 kvm's fragments

For kvm hypervisor guest on arm64, virtio mmio will be probed as the platform
device for 32 times even qemu hasn't created it. If we disable the virtio
mmio in kernel config, it will reduce 20-30ms for the boottime.

This disables virtio mmio in arm64 kvm's fragments.

Fixes: #985
Signed-off-by: Jia He <justin.he@arm.com>

justin-he commented:
> @justin-he - you only need to change the kernel version in one patch/commit :-) - and, it has moved on already, so you'll need to rebase/repush that.
> For the benchmark results - can you use the [metrics report tool](https://github.com/kata-containers/tests/tree/master/metrics/report) to get us a graph of the results over a number of runs - that way we can check if there is noise/variance, and not just a 'one off' win :-) thx!


Hi, it is hard to result significantly in metrics report because in theory it only reduce 20-30ms for boot time. But the whole boot time of each container will be 2-3s on arm64.
But in theory, it does no bad to boot time/memory footprint. The only disadvantage is qemu can't create any virtio mmio bus device (use virtio pci bus instead). IIUC, kata only use virtio pci device currently.

fidencio commented:
/test-arm

fidencio commented:
/test-ubuntu

fidencio reviewed (approved):
lgtm!

fidencio commented:
Folks, this PR has been stale and I'm closing it as we're walking towards the **final** 1.13.0 release of the kata-containers-1.x.
From now on, all the development will happen as part of https://github.com/kata-containers/kata-containers/ and fixes should be provided to that repo.

Thanks a lot for your contribution, we're sorry that for some reason it got stale, and we hope to see you around for new adventures in the kata-containers-2.x land.
